### ISSUE TITLE
KeypointRCNN Loss Index Error

### ISSUE THREAD

**Issue by byronverzmoter** [pytorch/vision#7750]
### 🐛 Describe the bug

 I am running into an IndexError when the `keypoint_loss` is being calculated:
```
        keypoint_loss = F.cross_entropy(keypoint_logits[valid], keypoint_targets[valid])
```
The issue is the `valid` indexes extend beyond the size of the `keypoint_logits` variable.
After a lot of debugging and investigating, the only pattern I notice is that the shape of the `keypoint_logits` is substantially larger when this error occurs (see below where keypoint_logits.shape is outputted):
```
N: 5, K: 4, H: 56, W: 56
N: 6, K: 4, H: 56, W: 56
N: 8, K: 4, H: 56, W: 56
N: 10, K: 4, H: 56, W: 56
N: 20, K: 4, H: 56, W: 56 <--- where the problem occurs
```
This does not happen on every image, and I have not been able to see a pattern on the input images and targets side that might suggest it is an error with the input data (but I know this could very well be where the issue is).

This is a custom dataset with only a single bounding box per image, 4 keypoints per bounding box, and a single label per bounding box. 

Here is an example of a target dictionary
```
{'boxes': tensor([[ 50.4433,  53.8947, 461.5564, 459.5478]]),
 'labels': tensor([3]),
 'keypoints': tensor([[ 50.4433, 170.1647,   1.0000],
         [442.2101,  53.8947,   1.0000],
         [461.5564, 343.2779,   1.0000],
         [ 69.7896, 459.5478,   1.0000]])}
```
The images are resized to (512,512) in the dataset.__getitem__() and I have ensured all target keypoints and bounding boxes fall within the image.

The shape of `keypoint_logits` is  `[20,  4, 56, 56]` when it is passed to the function, and `torch.Size([80, 3136])` after the line
```
    keypoint_logits = keypoint_logits.view(N * K, H * W)
```
The code to reproduce this would be too much to add everything, I think it would be easier if I could get some guidance on things to look for/extra info to provide here that would help.

### Versions

Collecting environment information...
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Enterprise
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.24.1
Libc version: N/A

Python version: 3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 05:59:45) [MSC v.1929 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19045-SP0
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: GeForce GTX 1650
Nvidia driver version: 462.30
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:

Revision=

Versions of relevant libraries:
[pip3] numpy==1.23.3
[pip3] torch==1.12.1
[pip3] torchaudio==0.12.1
[pip3] torchmetrics==1.0.0
[pip3] torchvision==0.13.1
[conda] blas                      2.116                       mkl    conda-forge
[conda] blas-devel                3.9.0              16_win64_mkl    conda-forge
[conda] cudatoolkit               11.3.1               h59b6b97_2  
[conda] libblas                   3.9.0              16_win64_mkl    conda-forge
[conda] libcblas                  3.9.0              16_win64_mkl    conda-forge
[conda] liblapack                 3.9.0              16_win64_mkl    conda-forge
[conda] liblapacke                3.9.0              16_win64_mkl    conda-forge
[conda] mkl                       2022.1.0           h6a75c08_874    conda-forge
[conda] mkl-devel                 2022.1.0           h57928b3_875    conda-forge
[conda] mkl-include               2022.1.0           h6a75c08_874    conda-forge
[conda] numpy                     1.23.3           py38h8503d5b_0    conda-forge
[conda] pytorch                   1.12.1          py3.8_cuda11.3_cudnn8_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.12.1               py38_cu113    pytorch
[conda] torchmetrics              1.0.0                    pypi_0    pypi
[conda] torchvision               0.13.1               py38_cu113    pytorch

**Comment by NicolasHug** [comment]
> The code to reproduce this would be too much to add everything, I think it would be easier if I could get some guidance on things to look for/extra info to provide here that would help.


Hmmmm I'd suggest looking into (almost) all the files in https://github.com/pytorch/vision/tree/main/references/detection and in particular what happens when keypoints are used / transformed there.

**Comment by byronverzmoter** [comment]
@NicolasHug  The keypoints targets that are passed don't appear to be an issue. I have gone through the reference implementation and I believe there is no issue directly with the targets or the images passed during training. 

After playing with the anchor generator parameters (scale and aspect ratios), it seems like modifying the `scale` has some impact on the section where I am getting an error, in fact, I get the exact same error at the first pass when I make the anchor generator scales large.

It seems that `keypoints_to_heatmap` outputs a square of shape `N x N`, but `keypoint_tensors` doesn't always have the first dimension of at least `N`. I'm not sure why this happens or what would be affecting the generation of the `keypoint_tensors`. Any suggestions here would be appreciated

**Comment by byronverzmoter** [comment]
After further investigation, it seems to have been an error on our end with the keypoints shape `[4,3]` instead of `[1,4,3]`. 

Could perhaps be useful to check that the first dimension of the `keypoints` matches the first dimension of `boxes` and raise an error indicating mismatch to the user?
